### ISSUE TITLE
Implement a Scoped Allocator to Fix Manual Memory Management

### ISSUE THREAD

**Issue by etcimon** [rejectedsoftware/vibe.d#938]
This is being rewritten because the solution is simpler than imagined.

**Comment by etcimon** [comment]
When `VibeManualMemoryManagement` is enabled, `readUntil` cannot free the underlying data. This currently seems to be a cause for a memory leak. The ideal `AllocAppender` allocator would be scoped with the lifetime of the task, which is why a Task Local instance of a `ScopedAllocator` would be ideal for this purpose. Also, a `DtorAllocator` can inherit the `ScopedAllocator` to accumulate object destructor pointers to destroy remaining instances cleanly.

For Bson, Json or DOM, serialization/deserialization (or any tree-like structure that doesn't free itself), a concept of `interface CollectedAllocator` could be used to replace `new`. This allocator only has an `alloc` abstract method. The `CollectedAllocator` will be passed with the constructor as with the `AllocAppender`.  The `ScopedAllocator` and `GCAllocator` can inherit this interface and thus be available to manage the memory used in this situation. This also opens the path for a thread-local or task-local GC in the future.

**Comment by etcimon** [comment]
@s-ludwig Do you see an issue with having an `Allocator m_heap` member in `Task` and using this as the default allocator if `fiber` is defined? This way, an allocator destructor can trash all the memory at once when the task ends, just like for process heaps.

**Comment by s-ludwig** [comment]
I think the first thing we should do (of course experimental changes can be done in parallel), is to start using scope where applicable. Anyone who wants to pass scoped objects to another task will then have to explicitly clone the object first. This would be the prerequisite to be able to do this kind of stuff in the official API. I'll start working on this as soon as the current dust is settled (still need to do a quick intermediate release to fix compilation on DMD 2.065 and GDC).

At least in the HTTP server, there is already such a scoped allocator ([`request_allocator`](https://github.com/rejectedsoftware/vibe.d/blob/10731f37585dde7567ce39c152365b1e197e323a/source/vibe/http/server.d#L1310)), just in a higher level place. So if we have a memory leak, that would mean that this allocator is currently not used everywhere during request processing. Or are you talking about the general case when `readUntil` is used together with an `AllocAppender` that uses the default allocator? It is supposed to always use `realloc`, so that only the final buffer needs to be freed - which can be done either using `.reset(AppenderResetMode.freeData)`, or by manually invoking the allocator. Any memory leak should thus be fixable from user code.

As for the general concept of a more general scoped allocation system, I think it would be very nice to have, but only with the proper type system support to make using it `@safe`. A few years ago there also was a project with a nice general approach to this problem (must be [this one](https://github.com/dsimcha/TempAlloc), [docs](http://cis.jhu.edu/~dsimcha/d/phobos/std_tempalloc.html)), but it never got into Phobos. Maybe it's time to try and revive that? We now have Andrei's proposal for an allocator interface and the lack of that was the main reason for rejecting it AFAIR.

**Comment by etcimon** [comment]
I was thinking of an approach that allowed the RedisClient or .. Any library to return string or array basic types without using the GC or taking the dst as a parameter or causing leaks. Having an RAII allocator on the overhead of tasks seems convenient for replacing the use of 'new' in Json, Bson, DOM objects where possible. In most cases, this data can be freed after the task is gone.

**Comment by etcimon** [comment]
I have a library up now where I refactored pretty much everything I need in terms of allocations in my projects. The relevant part would be here:

https://github.com/etcimon/memutils/blob/master/source/memutils/tests.d#L238

Would be interesting to have the option to replace many GC allocations with it, all that's needed is to link to it and call `destroyFiberPool()` in the `Task` destructor.

**Comment by etcimon** [comment]
Done

https://github.com/etcimon/spd
